lib/fileset/internal.nix

```diff
@@ -638,4 +638,66 @@ rec {
     else
       # In all other cases it's the rhs
       rhs;
+
+  # Compute the set difference between two file sets.
+  # The filesets must already be coerced and validated to be in the same filesystem root.
+  # Type: Fileset -> Fileset -> Fileset
+  _difference = fileset1: fileset2:
+    let
+      # The common base components prefix, e.g.
+      # (/foo/bar, /foo/bar/baz) -> /foo/bar
+      # (/foo/bar, /foo/baz) -> /foo
+      commonBaseComponentsLength =
+        # TODO: Have a `lib.lists.commonPrefixLength` function such that we don't need the list allocation from commonPrefix here
+        length (
+          commonPrefix
+            fileset1._internalBaseComponents
+            fileset2._internalBaseComponents
+        );
+
+      # The filesetTree's need to have the same base path in order to be able to compute the difference between them.
+      tree2 =
+        if commonBaseComponentsLength == length fileset1._internalBaseComponents then
+          # The common prefix is the same as the first path, so the second path is equal or longer.
+          # We need to shorten the second arguments tree to the same base path as the first one
```

Comment:
Not clear to me why.
What is `tree2` for?

Reply:
Same as the tree of the negative file set, but with the same base path as the positive file set, such that they are comparable and can be operated on.

E.g. if you have `difference /foo /foo/bar`, the common path would be `/foo`, meaning the first branch applies.

It then takes the negative file set, `(base: /foo/bar, tree: "directory")` in this case, and shortens the base to `/foo` instead, such that you get `(base: /foo, tree: { bar = "directory"; })`. `tree2` here only contains the tree part though, so `{ bar = "directory"; }`.

Now we have two trees, `"directory"` from the positive file set, and `{ bar = "directory"; }` from the negative file set, that have the same base `/foo`, so they can be operated on.

I'll put something like this in the code docs, and I'll improve the variable name, `negativeTreeWithPositiveBasePath` or so (maybe shorter :P)